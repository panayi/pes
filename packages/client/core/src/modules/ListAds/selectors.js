import * as R from 'ramda';
import { isPlainObj } from 'ramda-adjunct';
import { createSelector } from 'reselect';
import createCachedSelector from 're-reselect';
import propSelector from '@pesposa/core/src/utils/propSelector';
import id from '@pesposa/core/src/utils/id';
import * as constants from './constants';

const hitsSelector = propSelector('hits');
const hitSelector = propSelector('hit');
const sizeSelector = propSelector('size');
const fixedCardHeightSelector = propSelector('fixedCardHeight');
const hitIdSelector = R.compose(id, hitSelector);
const containerWidthSelector = propSelector('containerWidth');
const indexSelector = propSelector('index');
const columnHeightsSelector = propSelector('columnHeights');

const currentMinWidthSelector = createSelector(
  containerWidthSelector,
  sizeSelector,
  (containerWidth, size) =>
    R.find(R.gte(containerWidth), constants.MIN_WIDTHS(size)),
);

export const columnCountSelector = createSelector(
  sizeSelector,
  currentMinWidthSelector,
  R.compose(R.path(R.__, constants.COLUMN_COUNTS), R.unapply(R.identity)),
);

// columnWidthSelector :: { containerWidth } -> Number
export const columnWidthSelector = createSelector(
  containerWidthSelector,
  columnCountSelector,
  (containerWidth, columnCount) =>
    Math.max(
      0,
      (containerWidth + constants.GUTTER) / columnCount - constants.GUTTER,
    ),
);

// thumbnailSelector :: { hit } -> Thumbnail
//  Thumbnail = { url, height, dimensions }
export const thumbnailSelector = createCachedSelector(
  hitSelector,
  propSelector(['thumbnail']),
)(hitIdSelector);

export const thumbnailHeightSelector = createCachedSelector(
  fixedCardHeightSelector,
  thumbnailSelector,
  propSelector('columnWidth'),
  (fixedCardHeight, thumbnail, columnWidth) => {
    if (fixedCardHeight) {
      return fixedCardHeight - constants.CARD_CONTENT_HEIGHT;
    }

    const dimensions = R.prop('dimensions', thumbnail);
    return isPlainObj(dimensions)
      ? columnWidth / dimensions.width * dimensions.height
      : constants.DEFAULT_THUMBNAIL_HEIGHT;
  },
)(hitIdSelector);

// hitHeightSelector :: { hit, columnWidth } => Number
const hitHeightSelector = ({ hit, columnWidth }) => {
  if (hit.isPaidAd) {
    return hit.height + constants.PAID_AD_VERTICAL_PADDING * 2;
  }

  const thumbnailHeight = thumbnailHeightSelector({ hit, columnWidth });
  return Math.ceil(thumbnailHeight + constants.CARD_CONTENT_HEIGHT);
};

export const hitSizeAndPositionSelector = createSelector(
  indexSelector,
  hitsSelector,
  columnHeightsSelector,
  columnCountSelector,
  columnWidthSelector,
  fixedCardHeightSelector,
  (index, hits, columnHeights, columnCount, columnWidth, fixedCardHeight) => {
    const hit = hits[index];
    const minColumnHeight = R.reduce(
      R.min,
      Number.POSITIVE_INFINITY,
      columnHeights,
    );
    const column = R.findIndex(R.equals(minColumnHeight), columnHeights);
    const height =
      fixedCardHeight ||
      (hit && hitHeightSelector({ hit, columnWidth })) ||
      constants.DEFAULT_CARD_HEIGHT;
    const width = columnWidth;
    const x = column * (constants.GUTTER + width);
    const y = columnHeights[column] || 0;

    return {
      column,
      height,
      width,
      x,
      y,
    };
  },
);
