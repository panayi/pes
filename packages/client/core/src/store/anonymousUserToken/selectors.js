import * as R from 'ramda';
import * as constants from './constants';

// anonymousUserTokenSelector :: State -> Object | Nil
export const anonymousUserTokenSelector = R.prop(constants.ROOT_KEY);
