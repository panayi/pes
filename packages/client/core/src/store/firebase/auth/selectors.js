import * as R from 'ramda';
import { isNotNil } from 'ramda-adjunct';
import { createSelector } from 'reselect';
import { isLoaded } from 'react-redux-firebase';
import * as firebaseConfig from '@pesposa/core/src/config/firebase';
import decodeJwt from '@pesposa/core/src/utils/decodeJwt';

const FIREBASE_AUTH_PATH = [...firebaseConfig.FIREBASE_PATH, 'auth'];

// authErrorSelector :: State -> Object | Nil
export const authErrorSelector = R.path([
  ...firebaseConfig.FIREBASE_PATH,
  'authError',
]);

// firebaseSelector :: State -> Object
const firebaseAuthSelector = R.compose(
  R.defaultTo({}),
  R.path(FIREBASE_AUTH_PATH),
);

// isLoadedSelector :: State -> Boolean | Nil
const isLoadedSelector = createSelector(firebaseAuthSelector, isLoaded);

// isInitializingSelector :: State -> Boolean | Nil
const isInitializingSelector = R.path([
  ...firebaseConfig.FIREBASE_PATH,
  'isInitializing',
]);

export const uidSelector = createSelector(firebaseAuthSelector, R.prop('uid'));

// hasUidSelector :: State -> String | Nil
export const hasUidSelector = createSelector(uidSelector, isNotNil);

// isAnonymousSelector :: State -> Boolean | Nil
export const isAnonymousSelector = createSelector(
  firebaseAuthSelector,
  R.prop('isAnonymous'),
);

// isAuthenticatingSelector :: State -> Boolean
export const isAuthenticatingSelector = createSelector(
  isInitializingSelector,
  isLoadedSelector,
  R.useWith(R.or, [R.identity, R.not]),
);

// isAuthenticatedSelector :: State -> Boolean
export const isAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  isAnonymousSelector,
  (isAuthenticating, hasUid, isAnonymous) =>
    !isAuthenticating && hasUid && !isAnonymous,
);

// isNotAuthenticatedSelector :: State -> Boolean
// Equals to: !isAuthenticatingSelector && !hasUidSelector
export const isNotAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  isAnonymousSelector,
  (isAuthenticating, hasUid, isAnonymous) =>
    !isAuthenticating && (!hasUid || isAnonymous),
);

export const tokenSelector = createSelector(
  firebaseAuthSelector,
  R.path(['stsTokenManager', 'accessToken']),
);

export const isCurrentUserSelector = userIdSelector =>
  createSelector(
    isAuthenticatedSelector,
    uidSelector,
    userIdSelector,
    (isAuthenticated, currentUserId, userId) =>
      isAuthenticated && R.equals(currentUserId, userId),
  );

export const isAdminSelector = createSelector(
  tokenSelector,
  R.compose(R.prop('admin'), R.defaultTo({}), decodeJwt),
);
