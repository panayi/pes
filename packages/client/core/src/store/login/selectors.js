import * as R from 'ramda';
import { createSelector } from 'reselect';
import * as constants from './constants';

export const statusSelector = R.path(constants.STATUS_PATH);

const createFlagSelector = value =>
  createSelector(statusSelector, R.equals(value));

export const pendingSelector = createFlagSelector(
  constants.LOGIN_PENDING_STATUS,
);
export const succeededSelector = createFlagSelector(
  constants.LOGIN_SUCCESS_STATUS,
);
export const failedSelector = createFlagSelector(constants.LOGIN_FAIL_STATUS);
