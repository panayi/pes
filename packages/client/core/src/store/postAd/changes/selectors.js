import * as R from 'ramda';
import * as constants from './constants';

export const createdAdsSelector = R.path(constants.CREATED_ADS_PATH);

export const deletedAdsSelector = R.path(constants.DELETED_ADS_PATH);
