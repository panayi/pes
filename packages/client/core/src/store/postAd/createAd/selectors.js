import * as R from 'ramda';
import { createSelector } from 'reselect';
import * as constants from './constants';

const createAdStatusSelector = R.path(constants.CREATE_AD_STATUS_PATH);
export const createAdErrorSelector = R.path(constants.CREATE_AD_ERROR_PATH);

export const isCreateAdIdleSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_IDLE_STATE),
);

export const isCreateAdPendingSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_PENDING_STATE),
);

export const isCreateAdFailedSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_FAILED_STATE),
);

export const isCreateAdCompletedSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_COMPLETED_STATE),
);
