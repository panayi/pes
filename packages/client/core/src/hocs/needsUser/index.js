import * as R from 'ramda';
import { withRouter } from 'react-router-dom';
import { withProps } from 'recompose';
import { connectedRouterRedirect } from 'redux-auth-wrapper/history4/redirect';
import { selectors as authSelectors } from '../../store/firebase/auth';
import Spinner from '../../components/Spinner/Spinner';

const CenteredSpinner = withProps({
  centered: true,
})(Spinner);

const needsUser = options =>
  R.compose(
    withRouter,
    connectedRouterRedirect({
      redirectPath: '/',
      allowRedirectBack: false,
      authenticatedSelector: authSelectors.isAuthenticatedSelector,
      authenticatingSelector: authSelectors.isAuthenticatingSelector,
      AuthenticatingComponent: CenteredSpinner,
      ...R.defaultTo({}, options),
    }),
  );

export default needsUser;
