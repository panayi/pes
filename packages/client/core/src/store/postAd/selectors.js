import * as R from 'ramda';
import { isNilOrEmpty } from 'ramda-adjunct';
import { createSelector } from 'reselect';
import * as modelPaths from '@pesposa/core/src/config/modelPaths';
import propSelector from '@pesposa/core/src/utils/propSelector';
import { selectors as authSelectors } from '../firebase/auth';
import { selectors as userInfoSelectors } from '../userInfo';

// adImagesPathSelector :: Props -> Object | Nil
export const adImagesPathSelector = createSelector(
  propSelector('adId'),
  adId => `${modelPaths.ADS.string}/${adId}/props/images`,
);

// draftImagesPathSelector :: State -> Object | Nil
export const draftImagesPathSelector = createSelector(
  authSelectors.uidSelector,
  R.unless(R.isNil, uid => `${modelPaths.DRAFTS(uid).string}/images`),
);

export const newAdLocationSelector = createSelector(
  userInfoSelectors.locationSelector,
  location => {
    const countryCode = R.path(['address', 'country', 'cca2'], location);
    return isNilOrEmpty(countryCode)
      ? location
      : R.assocPath(['address', 'country'], countryCode, location);
  },
);

// adPropsForMixpanelSelector :: Ad -> Object
export const adPropsForMixpanelSelector = R.compose(
  R.evolve({
    location: R.prop('address'),
  }),
  R.pick(['title', 'category', 'location']),
);
