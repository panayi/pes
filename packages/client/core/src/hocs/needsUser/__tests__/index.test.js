import React from 'react';
import { mount } from 'enzyme';
import * as authSelectors from '../../../store/firebase/auth/selectors';
import needsUser from '../index';

describe('[HOC] needsUser', () => {
  authSelectors.isAuthenticatedSelector = jest.fn();
  authSelectors.isAuthenticatingSelector = jest.fn();

  const location = {
    pathname: '/some-user-protected-url',
    search: '',
    hash: '',
  };

  const SecretComponent = () => <div>i am a user</div>;
  const FailureComponent = () => <div>unauthorized</div>;
  const Hoc = needsUser({
    FailureComponent,
  })(SecretComponent);
  const { component } = withMockStore(<Hoc location={location} />);

  it('should render wrappedComponent correctly', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(true);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render wrappedComponent when isAuthenticatedSelector = true', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(true);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper.find(SecretComponent).exists()).toBe(true);
  });

  it('should render failureComponent, when isAuthenticatedSelector = false', () => {
    authSelectors.isAuthenticatedSelector.mockReturnValueOnce(false);
    authSelectors.isAuthenticatingSelector.mockReturnValueOnce(false);
    const wrapper = mount(component);
    expect(wrapper.find(SecretComponent).exists()).toBe(false);
    expect(wrapper.find(FailureComponent).exists()).toBe(true);
  });
});
