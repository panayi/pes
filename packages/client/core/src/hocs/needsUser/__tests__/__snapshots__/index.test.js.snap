// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[HOC] needsUser should render wrappedComponent correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "firebase": Object {
        "INTERNAL": Object {
          "ErrorFactory": [Function],
          "Promise": [Function],
          "createFirebaseNamespace": [Function],
          "createSubscribe": [Function],
          "deepExtend": [Function],
          "extendNamespace": [Function],
          "factories": Object {
            "auth": [Function],
            "database": [Function],
          },
          "node": Object {
            "XMLHttpRequest": [Function],
            "localStorage": Storage {},
            "sessionStorage": Storage {},
          },
          "registerService": [Function],
          "removeApp": [Function],
          "useAsService": [Function],
        },
        "Promise": [Function],
        "SDK_VERSION": "4.8.1",
        "User": [Function],
        "_": Object {
          "authUid": null,
          "config": Object {
            "apiKey": "asdf",
            "authDomain": "asdf",
            "databaseURL": "ws://127.0.1:5000",
            "enableRedirectHandling": false,
            "messagingSenderId": "asdf",
            "storageBucket": "asdf",
            "userProfile": "users",
          },
          "watchers": Object {},
        },
        "__esModule": true,
        "app": [Function],
        "apps": Array [],
        "auth": [Function],
        "database": [Function],
        "default": [Circular],
        "initializeApp": [Function],
      },
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(AuthWrapper(SecretComponent))
    location={
      Object {
        "hash": "",
        "pathname": "/some-user-protected-url",
        "search": "",
      }
    }
  >
    <AuthWrapper(SecretComponent)
      dispatch={[Function]}
      isAuthenticated={true}
      isAuthenticating={false}
      location={
        Object {
          "hash": "",
          "pathname": "/some-user-protected-url",
          "search": "",
        }
      }
    >
      <SecretComponent
        dispatch={[Function]}
        isAuthenticated={true}
        isAuthenticating={false}
        location={
          Object {
            "hash": "",
            "pathname": "/some-user-protected-url",
            "search": "",
          }
        }
      >
        <div>
          i am a user
        </div>
      </SecretComponent>
    </AuthWrapper(SecretComponent)>
  </Connect(AuthWrapper(SecretComponent))>
</Provider>
`;
